### DISCUSSION TITLE
Cannot commit: Not inside Git repository

### DISCUSSION THREAD

**Discussion by ebouna** [magit/magit#4973] [Support] [answered]
I cannot commit using Magit, whereas committing using Git in the bash works. I applied all possible remedies you suggest. None of them has worked so far. I would glad if you could give me some advice how to resolve the issue. I have tried reinstalling Magit and its dependencies, too. Thanks in advance.

Now, when I open emacs using the magit-emacs-Q-command and try to commit, I get:

`Error in post-command-hook (magit-section-post-command-hook): (args-out-of-range 145 145)`

The message in magit process says:
` 1 git … commit --
hint: Waiting for your editor to close the file...
Waiting for Emacs...
*ERROR*: Not inside Git repository
error: There was a problem with the editor '/usr/local/bin/emacsclient --socket-name=/run/user/1000/emacs/server'.
Please supply the message using either -m or -F option.`

Information on program/ package versions: 
`Magit 20230727.208 [>= 3.3.0.50-git], Transient 0.4.1, Git 2.41.0, Emacs 28.2, gnu/linux`

(Magit installed through Melpa, Emacs compiled from source code)

With-editor-debug output:

```
with-editor: /home/elias/.emacs.d/elpa/with-editor-20230711.1217/with-editor.el
emacs: /usr/local/bin/emacs (28.2)
system:
  system-type: gnu/linux
  system-configuration: x86_64-pc-linux-gnu
  system-configuration-options: --with-native-compilation --with-json --with-imagemagick --with-modules --with-mailutils --with-sound=yes
server:
  server-running-p: t
  server-process: #<process server>
  server-use-tcp: nil
  server-name: server
  server-socket-dir: /run/user/1000/emacs
    server
  server-auth-dir: ~/.emacs.d/server/
    WARNING: not an accessible directory
with-editor-emacsclient-executable:
 value:   /usr/local/bin/emacsclient (28.2)
 default: /usr/local/bin/emacsclient (28.2)
 funcall: /usr/local/bin/emacsclient (28.2)
path:
  $PATH:     (/home/elias/miniconda3/condabin /usr/local/texlive/2022/bin/x86_64-linux /home/elias/.local/bin /usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin /usr/games /usr/local/games /snap/bin /snap/bin /home/elias/.local/bin /home/elias/.local/bin)
  exec-path: (/home/elias/miniconda3/condabin /usr/local/texlive/2022/bin/x86_64-linux /home/elias/.local/bin /usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin /usr/games /usr/local/games /snap/bin /snap/bin /home/elias/.local/bin /home/elias/.local/bin /usr/local/libexec/emacs/28.2/x86_64-pc-linux-gnu)
  with-editor-emacsclient-path:
    /home/elias/miniconda3/condabin (t)
    /usr/local/texlive/2022/bin/x86_64-linux (t)
    /usr/local/sbin (t)
    /usr/local/bin (t)
      /usr/local/bin/emacsclient (28.2)
    /usr/sbin (t)
    /usr/bin (t)
    /sbin (usr/sbin)
    /bin (usr/bin)
    /usr/games (t)
    /usr/local/games (t)
    /snap/bin (t)
    /home/elias/.local/bin (t)
    /usr/local/libexec/emacs/28.2/x86_64-pc-linux-gnu (t)
```

**Comment by tarsius** [comment]
Do you (successfully) show the status buffer for that repository and then try to create a commit while that buffer is current? Have you tried with another repository?

>     Error in post-command-hook (magit-section-post-command-hook): (args-out-of-range 145 145)

Like all errors that occur during `post-command-hook`, this error should only happen once, because functions that cause issues are automatically removed from that hook. Do you still get the other error if you just repeat trying to create a commit?

**Reply by ebouna** [reply]
Thanks a lot for these tips. 
> Do you (successfully) show the status buffer for that repository and then try to create a commit while that buffer is current? 

Yes, I was able to successfully display the status buffer for the repository in question and then tried to create a commit while that buffer was current - which caused the issue described.

When I repeat, I still get the "Not inside Git repository" message. 

I have tried then this morning with freshly cloned repositories - cloned from other remote repositories, and I seem to have gained more insight into the issue: 
The problematic local repository, viz. the one my problems started with, is on a virtual machine (VirtualBox), and inside this virtual machine it is on a mounted folder, which is shared with the host system (windows 10) and which is in addition connected to OneDrive (a shared OneDrive folder) within the host system. 

When I create a new local repository at the same place where the first problematic local repository is, the same problems arise again. When I create a fresh repository outside the mounted folder on the virtual machine, Magit works as expected. 

So my question is now: How can I make Magit work in a mounted folder on a virtual machine, i.e. under the conditions mentioned? I have added user to the group the permissions of the folder belong to, but to no avail.

**Comment by tarsius** [comment]
> The problematic local repository, viz. the one my problems started with, is on a virtual machine (VirtualBox), and inside this virtual machine it is on a mounted folder, which is shared with the host system (windows 10) and which is in addition connected to OneDrive (a shared OneDrive folder) within the host system.

I am *guessing* that the mounted folder by itself causes the issue and that it is irrelevant that onedrive is also involved. But I could be totally wrong and think the next debugging step should be to try to reproduce with onedrive out of the picture.

> When I repeat, I still get the "Not inside Git repository" message.

That warning originates from one of the low-level functions signaling the `magit-outside-git-repo` error type. Grep for that, you will find all of that in the same vicinity. To learn what is calling which of those functions, enable debug-on-error. Unfortunately it possible we don't get a backtrace in this case.

My educated guess is that the problem is that `magit-toplevel` fails to account for yet another special-case. (That or some other function that is defined right above the error type.)

If you don't get a backtrace, I would recommend adding debugging statements to `magit-toplevel`, both to verify that the error is triggered here and to show intermediary values to determine where things start to take an unexpected turn.

**Reply by ebouna** [reply]
You guessed right: switching off OneDrive did not change anything. The same error occurred. 

After setting  the variable `debug-on-message` to "Git failed" I got the below backtrace (as I mentioned, OneDrive was turned off):

```
Debugger entered--Lisp error: "Git failed ... [Hit $ to see buffer magit-process:..."
  magit-process-finish(#<process git>)
  magit-process-sentinel(#<process git> "exited abnormally with code 1\n")
  sit-for(5)
  server-return-error(#<process server <4>> (magit-outside-git-repo . "/media/sf_OneDrive_-_Universitaet/git_repo/.g..."))
  server-execute(#<process server <4>> (("/media/sf_OneDrive_-_Universitaet/git_repo/.g...")) nil nil nil nil nil)
  #f(compiled-function () #<bytecode -0x1eca6a884bff61b>)()
  server-execute-continuation(#<process server <4>>)
  server-process-filter(#<process server <4>> "-dir /media/sf_OneDrive_-_Universitaet/git_re...")
```

Also I stepped through `magit-toplevel` with edebug after entering the directory `/media/sf_OneDrive_-_Universitaet/git_repo` and activating magit. But I did not get any error message. Instead the returned result was often `/media/sf_OneDrive_-_Universitaet/git_repo`. What seemed stranged to me, though, was that  `magit-toplevel` was called again and again. Unfortunately, I am not too experienced as far as edebug is concerned. So perhaps I missed something...

**Reply by tarsius** [reply]
(In markdown you should surround code-blocks with three backticks on either side.)

That backtrace unfortunately isn't very revealing. You could try visiting `magit-process.el` and then calling `eval-buffer`, same for `magit-git.el`. Maybe that would turn the `#f(compiled-function ...)` into something more useful.

> But I did not get any error message. Instead the returned result was often `/media/sf_OneDrive_-_Universitaet/git_repo`.

So the expected value, right?

> What seemed stranged to me, though, was that  `magit-toplevel` was called again and again.

It is called a lot during a buffer refresh, but only the first call should do any work. Subsequent calls during the same buffer refresh should use the cache, thanks to `magit--with-refresh-cache`. If the other code in this function is actually called multiple times (let's say more than twice), that might be due to the issue.

> I am not too experienced as far as edebug is concerned. So perhaps I missed something...

Maybe stick to debug statements to avoid heisenbugs.

Since you can refresh the status buffer, this is probably pointless, but you cannot reproduce the issue by calling `magit-toplevel` manually, right? And it returns the correct value? Oh, try inside the `.git` directory. Also, is the `.git` inside the working tree a directory or a plain file?

**Reply by ebouna** [reply]
> You could try visiting `magit-process.el` and then calling `eval-buffer`, same for `magit-git.el`. Maybe that would turn the `#f(compiled-function ...)` into something more useful.

I did as you suggested. This output from setting the variable `debug-on-message` to "Git failed" seems to be a bit more verbose (though `#f(compiled-function () `  remained ):

```
Debugger entered--Lisp error: "Git failed ... [Hit $ to see buffer magit-process:..."
  message("%s ... [%s buffer %s for details]" "Git failed" "Hit $ to see" "magit-process: git_repo")
  (cond (magit-process-raise-error (signal 'magit-git-error (list (format "%s (in %s)" msg default-dir)))) ((not (eq msg 'suppressed)) (if (buffer-live-p process-buf) (progn (save-current-buffer (set-buffer process-buf) (let* ((status-buf ...)) (if status-buf (save-current-buffer ... ...) nil))))) (message "%s ... [%s buffer %s for details]" msg (let* ((key (and t (and ... ...)))) (if key (format "Hit %s to see" (key-description key)) "See")) (buffer-name process-buf))))
  (let ((msg (magit-process-error-summary process-buf section))) (if magit-process-display-mode-line-error (magit-process-set-mode-line-error-status (or (magit-process-error-tooltip process-buf section) msg)) (magit-process-unset-mode-line default-dir)) (cond (magit-process-raise-error (signal 'magit-git-error (list (format "%s (in %s)" msg default-dir)))) ((not (eq msg 'suppressed)) (if (buffer-live-p process-buf) (progn (save-current-buffer (set-buffer process-buf) (let* (...) (if status-buf ... nil))))) (message "%s ... [%s buffer %s for details]" msg (let* ((key (and t ...))) (if key (format "Hit %s to see" (key-description key)) "See")) (buffer-name process-buf)))))
  (if (= arg 0) (magit-process-unset-mode-line default-dir) (let ((msg (magit-process-error-summary process-buf section))) (if magit-process-display-mode-line-error (magit-process-set-mode-line-error-status (or (magit-process-error-tooltip process-buf section) msg)) (magit-process-unset-mode-line default-dir)) (cond (magit-process-raise-error (signal 'magit-git-error (list (format "%s (in %s)" msg default-dir)))) ((not (eq msg 'suppressed)) (if (buffer-live-p process-buf) (progn (save-current-buffer (set-buffer process-buf) (let* ... ...)))) (message "%s ... [%s buffer %s for details]" msg (let* ((key ...)) (if key (format "Hit %s to see" ...) "See")) (buffer-name process-buf))))))
  magit-process-finish(#<process git>)
  (progn (setq event (substring event 0 -1)) (if (string-match "^finished" event) (progn (message (concat (capitalize (process-name process)) " finished")))) (magit-process-finish process) (if (eq process magit-this-process) (progn (setq magit-this-process nil))) (if (process-get process 'inhibit-refresh) nil (let ((command-buf (process-get process 'command-buf))) (if (buffer-live-p command-buf) (save-current-buffer (set-buffer command-buf) (magit-refresh)) (let ((temp-buffer (generate-new-buffer " *temp*" t))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn ... ...) (and ... ...))))))))
  (if (memq (process-status process) '(exit signal)) (progn (setq event (substring event 0 -1)) (if (string-match "^finished" event) (progn (message (concat (capitalize (process-name process)) " finished")))) (magit-process-finish process) (if (eq process magit-this-process) (progn (setq magit-this-process nil))) (if (process-get process 'inhibit-refresh) nil (let ((command-buf (process-get process 'command-buf))) (if (buffer-live-p command-buf) (save-current-buffer (set-buffer command-buf) (magit-refresh)) (let ((temp-buffer ...)) (save-current-buffer (set-buffer temp-buffer) (unwind-protect ... ...))))))))
  magit-process-sentinel(#<process git> "exited abnormally with code 1\n")
  recursive-edit()
  debug(error "Git failed ... [Hit $ to see buffer magit-process:...")
  message("%s ... [%s buffer %s for details]" "Git failed" "Hit $ to see" "magit-process: git_repo")
  (cond (magit-process-raise-error (signal 'magit-git-error (list (format "%s (in %s)" msg default-dir)))) ((not (eq msg 'suppressed)) (if (buffer-live-p process-buf) (progn (save-current-buffer (set-buffer process-buf) (let* ((status-buf ...)) (if status-buf (save-current-buffer ... ...) nil))))) (message "%s ... [%s buffer %s for details]" msg (let* ((key (and t (and ... ...)))) (if key (format "Hit %s to see" (key-description key)) "See")) (buffer-name process-buf))))
  (let ((msg (magit-process-error-summary process-buf section))) (if magit-process-display-mode-line-error (magit-process-set-mode-line-error-status (or (magit-process-error-tooltip process-buf section) msg)) (magit-process-unset-mode-line default-dir)) (cond (magit-process-raise-error (signal 'magit-git-error (list (format "%s (in %s)" msg default-dir)))) ((not (eq msg 'suppressed)) (if (buffer-live-p process-buf) (progn (save-current-buffer (set-buffer process-buf) (let* (...) (if status-buf ... nil))))) (message "%s ... [%s buffer %s for details]" msg (let* ((key (and t ...))) (if key (format "Hit %s to see" (key-description key)) "See")) (buffer-name process-buf)))))
  (if (= arg 0) (magit-process-unset-mode-line default-dir) (let ((msg (magit-process-error-summary process-buf section))) (if magit-process-display-mode-line-error (magit-process-set-mode-line-error-status (or (magit-process-error-tooltip process-buf section) msg)) (magit-process-unset-mode-line default-dir)) (cond (magit-process-raise-error (signal 'magit-git-error (list (format "%s (in %s)" msg default-dir)))) ((not (eq msg 'suppressed)) (if (buffer-live-p process-buf) (progn (save-current-buffer (set-buffer process-buf) (let* ... ...)))) (message "%s ... [%s buffer %s for details]" msg (let* ((key ...)) (if key (format "Hit %s to see" ...) "See")) (buffer-name process-buf))))))
  magit-process-finish(#<process git<1>>)
  (progn (setq event (substring event 0 -1)) (if (string-match "^finished" event) (progn (message (concat (capitalize (process-name process)) " finished")))) (magit-process-finish process) (if (eq process magit-this-process) (progn (setq magit-this-process nil))) (if (process-get process 'inhibit-refresh) nil (let ((command-buf (process-get process 'command-buf))) (if (buffer-live-p command-buf) (save-current-buffer (set-buffer command-buf) (magit-refresh)) (let ((temp-buffer (generate-new-buffer " *temp*" t))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn ... ...) (and ... ...))))))))
  (if (memq (process-status process) '(exit signal)) (progn (setq event (substring event 0 -1)) (if (string-match "^finished" event) (progn (message (concat (capitalize (process-name process)) " finished")))) (magit-process-finish process) (if (eq process magit-this-process) (progn (setq magit-this-process nil))) (if (process-get process 'inhibit-refresh) nil (let ((command-buf (process-get process 'command-buf))) (if (buffer-live-p command-buf) (save-current-buffer (set-buffer command-buf) (magit-refresh)) (let ((temp-buffer ...)) (save-current-buffer (set-buffer temp-buffer) (unwind-protect ... ...))))))))
  magit-process-sentinel(#<process git<1>> "exited abnormally with code 1\n")
  sit-for(5)
  server-return-error(#<process server <2>> (magit-outside-git-repo . "/media/sf_OneDrive_-_Universitaet/git_repo/.g..."))
  server-execute(#<process server <2>> (("/media/sf_OneDrive_-_Universitaet/git_repo/.g...")) nil nil nil nil nil)
  #f(compiled-function () #<bytecode 0x1ba2d4fe0b4009e5>)()
  server-execute-continuation(#<process server <2>>)
  server-process-filter(#<process server <2>> "-dir /media/sf_OneDrive_-_Universitaet/git_re...")

```

` /media/sf_OneDrive_-_Universitaet/git_repo` indeed should be the returned value.

When I go to the git repository, call `magit-status` and then evaluate `(magit-toplevel)` in the scratch buffer, the result is `nil`, same when I evaluate `(magit-toplevel "/media/sf_OneDrive_-_Universitaet/git_repo/.git")` in the scratch buffer. With `(magit-toplevel "/media/sf_OneDrive_-_Universitaet_Bern/git_repo/README.md")` it is `"/media/sf_OneDrive_-_Universitaet_Bern/git_repo` - as expected.

> Also, is the .git inside the working tree a directory or a plain file?

It is a directory.

BTW, what did you mean exactly by: 
> Oh, try inside the .git directory. 

?

**Reply by tarsius** [reply]
> This output

This seems weird. The error seems to not happen until `magit-process-finish`. So what I would expect to see is that the commit is actually created and only then something goes wrong, preventing the magit buffers from correctly reflecting the new state. But what probably happened is that the same issue happens twice, once when trying to run `git commit` and once afterwards when updating the state, and for some reason the first time around the debugger is not entered.

I am at a loss here. You will probably have to learn how these functions work yourself and incrementally narrow down your search for the code that doesn't work for you (but apparently everyone else).

> BTW, what did you mean exactly by:

> > Oh, try inside the .git directory.

`(let ((default-directory ".../.git")) (magit-toplevel))`

**Reply by cshansen1981** [reply]
FWIW I was having this problem as well. My setup runs WSL2 with a mounted ntfs partition that has the git checkout. Inside the .git folder it consistently returned nil when I did a `(magit-toplevel)` and `git commit` returned the "Not inside Git repository" message. Debugging a little it turned out Git actually printed:

```
fatal: detected dubious ownership in repository at 'PATH'
To add an exception for this directory, call:

        git config --global --add safe.directory [PATH]
```

It seems this was printed when `(magit-toplevel)` was running `git rev-parse --gitdir` internally. 

Running the `git config` command above fixed the issue. [PATH] is the location of the .git directory

Hope this helps

**Reply by axelknock** [reply]
I was also experiencing this issue just now. I also had no trouble with refreshing the magit buffer or committing from the command line. `(let ((default-directory ".../.git")) (magit-toplevel))` returned nil, and `(magit-toplevel)` returned the correct toplevel directory of the local repo. I also tried the command adding the directory to [safe], one with true path and one with the symlink path. Neither of these worked.
Investigation of the dubious ownership error yielded this issue:
https://stackoverflow.com/questions/72978485/git-submodule-update-failed-with-fatal-detected-dubious-ownership-in-reposit
I opted to add the following to my `.config/git/config` file:
```
[safe]
  directory = *
```
And this resolved the problem.
There is a submodule in my repo. Could this be the common thread?

**Comment by opensource-philosophy** [comment]
I'm afraid I have to confirm that I have the exact same problem, too. Additionally, I also get the "Not Inside Git Repository" error when trying to use magit-init. Here, the .git folder is created, but the error is shown anyway. When trying to commit, however, the commit command is not even executed. Lastly, some git repositories are not recognized as such although they contain a valid .git directory. When trying to call magit on a file in the git repo, I am prompted for a path and then asked whether a repo should be created in that path. When pressing "y", I also get "Not Inside Git Repository". When pressing "n", nothing happens. 

I first thought the issue was relevant to my .emacs configuration, but as a matter of fact, `emacs -q` yielded the exact same results. I am on Garuda Arch Linux.

**Reply by tarsius** [reply]
Exact some problem, including this part?:

> The problematic local repository, viz. the one my problems started with, is on a virtual machine (VirtualBox), and inside this virtual machine it is on a mounted folder, which is shared with the host system (windows 10) and which is in addition connected to OneDrive (a shared OneDrive folder) within the host system.

**Reply by opensource-philosophy** [reply]
My github folder containing the repos is on my (symlinked) storage partition and I do not use a virtual machine. What I mean by "exact same problem" is "same error message in the same situation".  The files are in the folder are synched by owncloud, though. It does not seem to matter, however: I copied one of the folders in a folder which is not synched at all and the problem persisted.

**Reply by tarsius** [reply]
Ebouna didn't mention any symlinks and is on Windows. So this is the same error in a *very different* situation.

But I am afraid you are in a similar situation, in that you have to learn how these functions work yourself and incrementally narrow down your search for the code that doesn't work for you.

**Comment by tarsius** [comment, marked as answer]
Setting the new `git-commit-cd-to-toplevel` to `t` might help with this. Also see its docstring and maybe #5100.

**Reply by divegeek** [reply]
This solved my incarnation of this problem.  I'm not using a VM or remote repository or anything like that, but I've found that for the last year or so I can't use magit to commit to repos in an Android checkout (Android uses the `repo` tool to manage a large set of git repositories).

I don't really understand the problem, or why this solves it. Maybe I'll dig into it when I get some spare time (ha!).
